### PR TITLE
Use @cpuguy83's patch and add a global cleanup lock

### PR DESCRIPTION
**Pull request by etamponi**
I got stuck in #5618 and #17443 for a week without finding a good solution, except
for the great @cpuguy83 's PR #23178 that unlocked at least the readers (ps, inspect)
but could not avoid NMI watchdog soft lockups nor unregister_netdevice errors.

I found the issue #19758 and https://github.com/docker/docker/issues/19758#issuecomment-182305310
 gave me the idea: since it only occours for concurrent cleanups, why not
putting a Global Cleanup Lock in place? (I come from a Python background, does it show? 😄  )

It works well, I have used a modified version of the docker stress tool that you can find
at https://github.com/etamponi/docker-stress . I used this configuration:

```
docker-stress --containers 10000 --concurrent 200 --kill 3s busybox
```

with the `stress.json` file that I put in that repo. I tried several times, and no issues of
sort happened. Using the same command on v1.11.2, or v1.10 and any kernel from 3.16 to 4.5 causes
NMI watchdog lockups and/or unregister_netdevice errors.

Of course this is a POC patch that shows where the issue is. From here we can try to add finer
locks, only where they are needed (perhaps just around the `releaseNetwork` call). But until the
kernel data race is fixed I fear we cannot get anything less than a "global" lock somewhere.

Beautiful animal: an hummigbird :)

screen shot 2016-06-03 at 17 20 08

Signed-off-by: Emanuele Tamponi emanuele.tamponi@gmail.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
I tried to do something similar but in the bridge driver itself around endpoint creation and deletion, which unfortunately didn't work.... I'll test this out. We should try to get this as deep in the stack as possible and only when hairpin mode is enabled.
I suspect if this is indeed working (where as in the bridge it was failing) it may be due to `Cleanup` being a significantly slower operation, meaning less chances of triggering the kernel race.

**Comment by etamponi** [comment]
@cpuguy83 , I tried to just disable hairpin mode, but without this change I would still hit NMI watchdog soft lockups and unregister_netdevice data races... If you can show me the file in the kernel in which you worked, I can try too and look around 😄

**Comment by cpuguy83** [comment]
hairpin mode is enabled by setting `--userland-proxy=false` on the daemon.

**Comment by etamponi** [comment]
Yeah I know, I tried with both `--userland-proxy=false` and without it, I was getting the same issues

**Comment by etamponi** [comment]
I have run some additional tests, to verify exactly all the failure scenarios.

This is what happens on Ubuntu Xenial Xerus, with Docker 1.11.2 (so _without_ this patch):

    aufs     userland-proxy=true   ->  docker runs well
    overlay  userland-proxy=true   ->  NMI watchdog soft lockups
    aufs     userland-proxy=false  ->  unregister_netdevice locks
    overlay  userland-proxy=false  ->  NMI watchdog soft lockups

I don't know the origin of the NMI watchdog soft lockups. The stack trace seems to suggest it's related to a `show_mountinfo` call, but I have no additional insights on it. I originally thought it was a different symptom of the same problem with the `veth` driver. 

On the other hand, I can now confirm that the `unregister_netdevice` locks only show up when you use hairpin mode (`--userland-proxy=false`).

This PR seems to address both the NMI watchdog soft lockups _*and*_ the unregister_netdevice locks. I think that the reason why this is effective is what cpuguy83 already described, that is that we are making things significantly slower, and this way we don't incur in kernel data races.

Anyway, if you want to use aufs and you can go with the userland proxy, stock Docker or Docker patched with #23178 work very well.

**Comment by etamponi** [comment]
More updates: both `unregister_netdevice: waiting for <dev> to become free` and `NMI watchdog: BUG: soft lockup` errors went away for me when I disabled IPv6 *_completely_*: I used `--ipv6=false` in Docker and `ipv6.disable=1` in grub as described in https://wiki.debian.org/DebianIPv6#How_to_turn_off_IPv6

**TL;DR**: I disable IPv6, I can use `overlay`, hairpin mode, or both, without any issue locks (but PR #23178 is definitely still needed to speed up reader tasks, in particular with `overlay`).

**Comment by edevil** [comment]
@etamponi I'm also having Docker freezes with the ```unregister_netdevice: waiting for veth846b1dc to become free. Usage count = 1``` message. Can you confirm that this problem goes away when we disable IPv6 on the node?

I'm running Kubernetes 1.2.4 on CoreOS beta (1010.4.0)/Flannel/Azure. Thanks.

**Comment by etamponi** [comment]
I am 100% sure that it worked for us on every machine that we run (22 EC2
instances).

We run an Ubuntu 16.04 AMI (patched kernel because it has an outstanding
bug with user-remap, but should be fixed next week), and just Docker
Engine, no Kubernetes.

Tested also on debian 8.4, and it works fine there too
Il 09 giu 2016 1:28 AM, "André Cruz" <notifications@github.com> ha scritto:

**Comment by devec0** [comment]
@edevil We (@etamponi and I) found that this specific issue went away if IPv6 was disabled via kernel command line (`ipv6.disable = 1`) or by compiling a kernel with IPV6 explicitly disabled (`CONFIG_IPV6` disabled).

We also found that the aufs storage driver is more stable for us than overlay or devicemapper, as even with IPv6 disabled, we would still get NMI soft lockups. devicemapper is also not performant for our use case.

With IPV6 disabled (both in kernel, and on the docker daemon commandline) and AUFS enabled, and with user proxy disabled on the docker daemon commandline (the user proxy chews a fair bit of RAM, so we prefer not to use it) on the latest Ubuntu Xenial kernel, we are not able to reproduce this issue :tada: 

Note also, that the above combination works with the stock docker daemon, not the modified version described in PR #23178. There's definitely still some work to be done on locking, though. I haven't had much of a chance to dig into the netlink code myself, so @etamponi can speak more to that.

**Comment by edevil** [comment]
In Azure I don't think you can pass kernel flags, what I can do is disable IPv6 on all interfaces at bootup via sysctl:

    net.ipv6.conf.all.disable_ipv6 = 1

Did you try this route? Also, did you try just disabling IPv6 in the docker daemon or just the kernel? Maybe we don't need both.

**Comment by devec0** [comment]
Yep, we tried that @edevil but unfortunately didn't have the same effect. 

For reference, we tried:
```
net.ipv6.conf.all.disable_ipv6=1 #disables for current interfaces
net.ipv6.conf.default.disable_ipv6=1 #disables for future interfaces
```

The above two sysctls did not disable IPv6, rather disable it on an interface level in the host. Because IPV6 is still enabled in the host kernel, it seems the containers would still try to allocate/discover IPv6 addresses, which still led to the `refcnt` issues mentioned above.

The only thing that worked was disabling on the kernel commandline or in the kernel configuration itself and building a custom kernel. I'm sure that if IPv6 support is not present in the kernel, then the `--ipv6=false` on the docker daemon commandline is redundant, but we didn't spend much time playing around with the option to be honest, once we got things working :smile: 

As an aside, on Azure, you're right - I couldn't find any documentation on how to edit kernel flags or even run a different kernel. You might try editing your grub config and seeing where `update-grub` gets you on a test instance. We're using AWS, where we use `pvgrub` on HVM machines so we can manage grub configuration like a regular VM or baremetal host. That's not a recommendation to use AWS, just an explanation of how it's done on AWS in case it helps you out on Azure.

**Comment by edevil** [comment]
FYI, I managed to boot the vm on Azure with ```ipv6.disable=1```, but it did not have network access. It seems IPv6 is needed in this case. :(

**Comment by etamponi** [comment]
The alternative is to enable the userland proxy and use AUFS. That will work also with ipv6 enabled

**Comment by edevil** [comment]
@etamponi Ah, great. I had forgotten about that option. Thanks.

**Comment by cpuguy83** [comment]
I think we should take this in two steps.

First would be fixing the locking on containers:
1. We need a way to safely access container fields
2. We need a safe way to make state changes, without blocking 1.

I actually closed my original PR since I was not happy with it, but maybe we can iterate here.

**Comment by sethp-jive** [comment]
I would suggest using [this commit](https://github.com/sethp-jive/docker/commit/9b5b2768e512a78c6ba28a68986b78665cb55a8c) as a starting point – that change reduces the critical section of the container lock so as not to include the netlink call, vs. adding multiple overlapping critical sections that do include the operation we know to be dangerous. I didn't submit it for consideration because it was only a point solution to the `netdevice_unregister` problem and did little to address your 1 or 2 generally, but I'm happy to add "signed-off-by" and click "create PR" if you'd like (or, feel free to just take the code).

For that general problem, broadly it seems like we'd want to implement the equivalent of "READ_COMMITTED" where we always have read access to the last unmodified Container object even while updates are in progress. Some difficulty-increasing factors:

1. There are lots of concurrent writers and readers
2. Docker's writes are basically designed to perform side-effects (which complicates conflict resolution)
3. Many writes touch lots of different fields on the Container (roughly: there's no clear way ahead of time to tell which write operations can safely operate concurrently)

For myself, I don't know much of the internals of docker, but I suspect that knowing and enumerating all the ways a Container is written to is a challenge at this point.

This change to me feels like it started down the road of introducing a non-exclusive reader/writer lock (non-exclusive meaning readers are fine concurrent with writers). Maybe we can formalize that by keeping a different copy of all the Containers that the readers consume, and having a goproc that syncs them up "every once in a while"? Alternatively, is it feasible at this point to find all of the Container writers and change their behavior, or should we focus solutions on the readers?

**Comment by manoj0077** [comment]
@sethp-jive I tried the commit you suggested above, I am seeing issues with restart of container with below errors 
"ERRO[0177] failed to cleanup up stale network sandbox for container YYY another sandbox delete in progress"
Cannot restart container YYY: service endpoint with name trusting_ardinghelli already exists.....
Above is because while cleanup is in progress, we are trying to start the container and sb.delete() is throwing error and endpoints deletion is not happening ?

Any clue on how to wait till cleanup operation before starting the container ?

**Comment by manoj0077** [comment]
@sethp-jive I tried the commit you suggested above, I am seeing issues with restart of container with below errors

"ERRO[0177] failed to cleanup up stale network sandbox for container YYY another sandbox delete in progress"
Cannot restart container YYY: service endpoint with name trusting_ardinghelli already exists.....

Above is because while cleanup is in progress, we are trying to start the container and sb.delete() is throwing error saying "it is already in progress" and before endpoints deletion happened we are creating similar endpoints....

Any clue on how to wait till cleanup operation before starting the container ?

**Comment by manoj0077** [comment]
@sethp-jive I tried the commit you suggested above, I am seeing issues with restart of container with below errors

"ERRO[0177] failed to cleanup up stale network sandbox for container YYY another sandbox delete in progress"
Cannot restart container YYY: service endpoint with name trusting_ardinghelli already exists.....

Above is because while cleanup is in progress, we are trying to start the container and sb.delete() is throwing error saying "it is already in progress" and before endpoints deletion happened we are creating similar endpoints....

Any clue on how to wait till cleanup operation before starting the container ?

**Comment by CVTJNII** [comment]
So this will resolve 19758?  If so very :+1:

**Comment by vdemeester** [comment]
@cpuguy83 what is the status here ? 👼

**Comment by cpuguy83** [comment]
Thanks for working on this patch.
I don't think this is the route we want to go, though.
I am currently working on removing the container lock all-together.

In addition it sounds like the kernel bug which is causing the daemon to lockup is fixed and scheduled to be ported to stable...looks like Canonical is already working on backporting this to 16.04 as well... so definitely things are looking up here.

Thanks again!

**Comment by sogabe** [comment]
Hello,

Can anyone please give us a reference of the patch in Ubuntu16.04 kernel?

> In addition it sounds like the kernel bug which is causing the daemon to lockup is fixed and scheduled to be ported to stable...looks like Canonical is already working on backporting this to 16.04 as well... so definitely things are looking up here.
